crates/shared/src/price_estimation/native/coingecko.rs

```diff
@@ -158,19 +178,36 @@ impl CoinGecko {
         token: &Token,
         denominator_price_eth: Decimal,
         prices: &HashMap<Token, f64>,
+        infos: &HashMap<Token, TokenInfo>,
+        denominator: &Denominator,
     ) -> NativePriceEstimateResult {
+        let token_decimals = infos
+            .get(token)
+            .and_then(|i| i.decimals)
+            .with_context(|| format!("missing decimals: {token:?}"))?;
+
         let token_price_eth: Decimal = prices
             .get(token)
             .cloned()
             .ok_or(PriceEstimationError::NoLiquidity)?
             .try_into()
             .context("failed to convert price to decimal")?;
 
-        Ok(token_price_eth
+        // When the quoted token and the denominator have different number of decimals
+        // the computed price effectively needs to be shifted by the difference.
+        let adjustment = Decimal::new(10, 0)
+            .checked_powi(i64::from(denominator.decimals) - i64::from(token_decimals))
+            .context("price adjustment overflows Decimal")?;
+
+        let price_denominated_in_token = token_price_eth
             .checked_div(denominator_price_eth)
             .context("division by zero")?
+            .checked_mul(adjustment)
```

Comment:
Do we need to do the adjustment before or after the division to not lose precision?

Reply:
I don't think so. Since this is not using a float under the hood but a `(base: i64, scale: u32)` the order of operations should not matter AFAIU.